kv: add TestTxnReadWithinUncertaintyInterval

nvb opened the pull request:
Extracted from #72121, because that PR is on-hold.

The first commit adds a new test called `TestTxnReadWithinUncertaintyInterval` which tests cases where a transaction observes a committed value below its global uncertainty limit while performing a read. In one variant, the transaction does not have an observed timestamp, so it hits a ReadWithinUncertaintyIntervalError. In a second variant, it does have an observed timestamp, so it avoids the error.

This is very old functionality, but I could not find a functional test like this in `pkg/kv/kvserver`, so I figured I'd add one. We have tests below this level (in pkg/storage) and above this level (jepsen, I hope elsewhere as well), but I couldn't find a similar test operating at this exact level.

The second commit adds an additional variant of this test which exercises a case where a transaction observes a committed value in its uncertainty interval that was written under a previous leaseholder. In the test, the transaction does collect an observed timestamp from the KV node that eventually serves the read, but this observed timestamp is not allowed to be used to constrain its local uncertainty limit below the lease start time and avoid the uncertainty error. As a result, it observes the committed value in its uncertain future and receives a ReadWithinUncertaintyIntervalError, which avoids a stale read.

nvb committed:
kv: add TestTxnReadWithinUncertaintyInterval

This commit adds a new test called TestTxnReadWithinUncertaintyInterval which
tests cases where a transaction observes a committed value below its global
uncertainty limit while performing a read. In one variant, the transaction does
not have an observed timestamp, so it hits a ReadWithinUncertaintyIntervalError.
In a second variant, it does have an observed timestamp, so it avoids the error.

This is very old functionality, but I could not find a functional test like this
in `pkg/kv/kvserver`, so I figured I'd add one. We have tests below this level
(in pkg/storage) and above this level (jepsen, I hope elsewhere as well), but I
couldn't find a similar test operating at this exact level.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/73725)
<!-- Reviewable:end -->

nvb committed:
kv: add TestTxnReadWithinUncertaintyIntervalAfterLeaseTransfer

This commit adds a new test called
`TestTxnReadWithinUncertaintyIntervalAfterLeaseTransfer`, which is a variant of
the `TestTxnReadWithinUncertaintyInterval` added in the prior commit. This test
exercises a case where a transaction observes a committed value in its
uncertainty interval that was written under a previous leaseholder. In the test,
the transaction does collect an observed timestamp from the KV node that
eventually serves the read, but this observed timestamp is not allowed to be
used to constrain its local uncertainty limit below the lease start time and
avoid the uncertainty error. As a result, it observes the committed value in its
uncertain future and receives a ReadWithinUncertaintyIntervalError, which avoids
a stale read.

The test is fairly similar to `TestRangeLocalUncertaintyLimitAfterNewLease`, but
written slightly differently and in the style that I intend to use for other
upcoming changes that touch this logic.

The test fails if I comment out this line:
https://github.com/cockroachdb/cockroach/blob/22df0a6658a927e7b65dafbdfc9d790500791993/pkg/kv/kvserver/uncertainty/compute.go#L90

lidorcarmel reviewed (approved):
Nice!

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73725#-:-Mqm0wzY4Z0VsQz7VWNp:b-83sjoh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)



<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
TFTR!

bors r=lidorcarmel
